### PR TITLE
ESX:virt-who config API cases support for SCA enabled

### PR DESCRIPTION
**Pull request by yanpliu**
Realized ESX:virt-who config API cases support for SCA enabled
API cases:PASS
```
(robottelo_vv) [root@dell-per740-68-vm-04 robottelo]# pytest ./tests/foreman/virtwho/api/test_esx_sca.py
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0
shared_function enabled - OFF - scope:  - storage: file
rootdir: /root/gitrepo/robottelo, configfile: pyproject.toml
plugins: xdist-3.1.0, services-2.2.1, mock-3.10.0, reportportal-5.1.3, ibutsu-2.2.4, cov-3.0.0
collected 10 items / 10 deselected / 0 selected                                                                                                                                                                   

tests/foreman/virtwho/api/test_esx_sca.py ...........                                                                                                                                                       [100%]
================================================================================================ warnings summary =================================================================================================
../../../home/yanpliu/data/virtualenv_310/robottelo_vv/lib/python3.10/site-packages/_pytest/config/__init__.py:747
  /home/yanpliu/data/virtualenv_310/robottelo_vv/lib/python3.10/site-packages/_pytest/config/__init__.py:747: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: pytest_fixtures.component.satellite_auth
    self.import_plugin(import_spec)

tests/foreman/virtwho/api/test_esx_sca.py: 83 warnings
  /home/yanpliu/data/virtualenv_310/robottelo_vv/lib/python3.10/site-packages/urllib3/connectionpool.py:1045: InsecureRequestWarning: Unverified HTTPS request is being made to host 'dell-per740-68-vm-07.lab.eng.pek2.redhat.com'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

====================================================================== 11 passed, 11 deselected, 84 warnings in 1113.16s (0:18:33) =======================================================================
```

### CODE REVIEW

File: tests/foreman/virtwho/api/test_esx_sca.py [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+    }
+    return form
+
+
+@pytest.fixture()
+def virtwho_config(form_data, target_sat):
+    virtwho_config = target_sat.api.VirtWhoConfig(**form_data).create()
+    yield virtwho_config
+    virtwho_config.delete()
+    assert not target_sat.api.VirtWhoConfig().search(query={'search': f"name={form_data['name']}"})
```
Comment (JacobCallahan) [inline tests/foreman/virtwho/api/test_esx_sca.py:56]: what is the value of this assertion, given that it would happen after test execution?

File: tests/foreman/virtwho/api/test_esx_sca.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+        :id: 61eeef42-12a0-4b92-9b87-c409b2507052
+
+        :expectedresults: debug option can be updated.
+
+        :CaseLevel: Integration
+
+        :CaseImportance: Medium
+        """
+        options = {'true': '1', 'false': '0', '1': '1', '0': '0'}
+        for key, value in sorted(options.items(), key=lambda item: item[0]):
```
Comment (JacobCallahan) [inline tests/foreman/virtwho/api/test_esx_sca.py:113]: instead of sorting a static dictionary at runtime, try instead to define the dictionary in the order you want. Then you just need to iterate over an already sorted dictionary.

File: tests/foreman/virtwho/api/test_esx_sca.py [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+            '60': '3600',
+            '120': '7200',
+            '240': '14400',
+            '480': '28800',
+            '720': '43200',
+            '1440': '86400',
+            '2880': '172800',
+            '4320': '259200',
+        }
+        for key, value in sorted(options.items(), key=lambda item: int(item[0])):
```
Comment (JacobCallahan) [inline tests/foreman/virtwho/api/test_esx_sca.py:148]: same comment here and the sort doesn't appear to be necessary since it is already in the correct order.
```suggestion
        for key, value in options.items():
```

File: tests/foreman/virtwho/api/test_esx_sca.py [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+        :id: fb1ebabe-7522-45a3-8e27-05101c50aabe
+
+        :expectedresults: hypervisor_id option can be updated.
+
+        :CaseLevel: Integration
+
+        :CaseImportance: Medium
+        """
+        values = ['uuid', 'hostname']
+        for value in values:
```
Comment (JacobCallahan) [inline tests/foreman/virtwho/api/test_esx_sca.py:174]: we can simplify this a bit
```suggestion
        for value in ['uuid', 'hostname']:
```

File: tests/foreman/virtwho/api/test_esx_sca.py [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+            virtwho_config.update(['hypervisor_id'])
+            config_file = get_configure_file(virtwho_config.id)
+            command = get_configure_command(virtwho_config.id, module_sca_manifest_org.name)
+            deploy_configure_by_command(
+                command, form_data['hypervisor_type'], org=module_sca_manifest_org.label
+            )
+            assert get_configure_option('hypervisor_id', config_file) == value
+
+    @pytest.mark.tier2
+    def test_positive_filter_option(self, module_sca_manifest_org, form_data, target_sat):
```
Comment (JacobCallahan) [inline tests/foreman/virtwho/api/test_esx_sca.py:185]: this is one impressive test! do you think it would be possible to break it up with parametrization? I see a lot of repeated logic which typically indicates a good candidate for splitting/parametrizing.

File: tests/foreman/virtwho/cli/test_esx_sca.py [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+        :id: 995a6709-e839-4198-89db-37cde8fd0a7b
+
+        :expectedresults: hypervisor_id option can be updated.
+
+        :CaseLevel: Integration
+
+        :CaseImportance: Medium
+        """
+        values = ['uuid', 'hostname']
+        for value in values:
```
Comment (JacobCallahan) [inline tests/foreman/virtwho/cli/test_esx_sca.py:113]: same suggestion
```suggestion
        for value in ['uuid', 'hostname']:
```

File: tests/foreman/virtwho/cli/test_esx_sca.py [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+
+        :expectedresults: debug option can be updated.
+
+        :CaseLevel: Integration
+
+        :CaseImportance: Medium
+        """
+        assert virtwho_config['name'] == form_data['name']
+        options = {'true': '1', 'false': '0', 'yes': '1', 'no': '0'}
+        for key, value in sorted(options.items(), key=lambda item: item[0]):
```
Comment (JacobCallahan) [inline tests/foreman/virtwho/cli/test_esx_sca.py:142]: same comment about sorting the dictionary on the definition.

File: tests/foreman/virtwho/cli/test_esx_sca.py [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+        for key, value in sorted(options.items(), key=lambda item: int(item[0])):
+            target_sat.cli.VirtWhoConfig.update({'id': virtwho_config['id'], 'interval': key})
+            command = get_configure_command(virtwho_config['id'], module_sca_manifest_org.name)
+            deploy_configure_by_command(
+                command, form_data['hypervisor-type'], org=module_sca_manifest_org.label
+            )
+            assert get_configure_option('interval', ETC_VIRTWHO_CONFIG) == value
+
+    @pytest.mark.tier2
+    def test_positive_filter_option(
```
Comment (JacobCallahan) [inline tests/foreman/virtwho/cli/test_esx_sca.py:208]: Same question about breaking this into parametrized tests.

**Review by JacobCallahan — CHANGES_REQUESTED** [review]
Very impressive! Just a few suggestions and questions.

File: tests/foreman/virtwho/api/test_esx_sca.py (hunk 1 above)
Comment (yanpliu) [reply tests/foreman/virtwho/api/test_esx_sca.py:56]: > what is the value of this assertion, given that it would happen after test execution?

Hello Jack, we want to check if the virt-who config has been deleted, and not exist

File: tests/foreman/virtwho/api/test_esx_sca.py (hunk 2 above)
Comment (yanpliu) [reply tests/foreman/virtwho/api/test_esx_sca.py]: > instead of sorting a static dictionary at runtime, try instead to define the dictionary in the order you want. Then you just need to iterate over an already sorted dictionary.

@JacobCallahan Have realdy updated the code with order 
```
options = {'0': '0', '1': '1', 'false': '0', 'true': '1'}
        for key, value in options.items():
```

File: tests/foreman/virtwho/cli/test_esx_sca.py (hunk 8 above)
Comment (yanpliu) [reply tests/foreman/virtwho/cli/test_esx_sca.py:209]: > Same question about breaking this into parametrized tests.

@JacobCallahan Have already parametrized tests.
```
(robottelo_vv) [root@dell-per740-68-vm-04 robottelo_20230307]# pytest  ./tests/foreman/virtwho/cli/test_esx_sca.py  -k test_positive_filter_option  --disable-pytest-warnings -q
....                                                                                                                                                                                                        [100%]
4 passed, 12 deselected, 7 warnings in 320.71s (0:05:20)

```

File: tests/foreman/virtwho/api/test_esx_sca.py (hunk 5 above)
Comment (yanpliu) [reply tests/foreman/virtwho/api/test_esx_sca.py]: > this is one impressive test! do you think it would be possible to break it up with parametrization? I see a lot of repeated logic which typically indicates a good candidate for splitting/parametrizing.

@JacobCallahan Have already parametrized tests.
```
(robottelo_vv) [root@dell-per740-68-vm-04 robottelo_20230307]# pytest  ./tests/foreman/virtwho/api/test_esx_sca.py  -k test_positive_filter_option  --disable-pytest-warnings -q
....                                                                                                                                                                                                        [100%]
4 passed, 12 deselected, 23 warnings in 234.93s (0:03:54)
```

**Review by JacobCallahan — APPROVED** [review]
ACK, thanks for making the adjustments!

**Comment by JacobCallahan** [comment]
@yanpliu looks like you have a merge conflict now

File: tests/foreman/virtwho/api/test_esx.py [hunk 9]
```diff
@@ -177,7 +177,7 @@ def test_positive_debug_option(self, default_org, form_data, virtwho_config, tar
 
         /foreman_virt_who_configure/api/v2/configs/:id"
 
-        :id: be395108-3944-4a04-bee4-6bac3fa03a19
+        :id: 24bf4c5b-cc7a-40ac-9079-e59c73984081
```
Comment (Gauravtalreja1) [inline tests/foreman/virtwho/api/test_esx.py:180]: Could you revert this change for the test case ids? it seems irrelevant.

**Review by Gauravtalreja1 — CHANGES_REQUESTED** [review]
